### PR TITLE
Stabilized `#[repr(align(x))]` attribute (RFC 1358)

### PR DESCRIPTION
**Pull request by bitshifter**
Stabilzed `#[repr(align(x))]` with attr_literal syntax as proposed by @eddyb https://github.com/rust-lang/rust/issues/33626#issuecomment-348467804

### CODE REVIEW

**Comment by eddyb** [comment]
cc @jseyfried @alexcrichton

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -533,6 +530,8 @@ declare_features! (
     // Allows the sysV64 ABI to be specified on all platforms
     // instead of just the platforms on which it is the C ABI
     (accepted, abi_sysv64, "1.24.0", Some(36167)),
+    // Allows `repr(align(u16))` struct attribute (RFC 1358)
```
Comment (eddyb) [inline src/libsyntax/feature_gate.rs]: This should say `16` not `u16`.

Comment (bitshifter) [reply src/libsyntax/feature_gate.rs]: Ah yes, my intention was to specify the type of the alignment parameter, but it's totally unclear here and no longer u16 anyway.

**Comment by alexcrichton** [comment]
cc @rust-lang/lang, would one of y'all like to initiate the fcpbot stabilization?

**Comment by cramertj** [comment]
@rfcbot fcp merge

**Comment by carols10cents** [comment]
@aturon @nrc waiting for your ticky boxes [here](https://github.com/rust-lang/rust/pull/47006#issuecomment-355344754)!

**Comment by bitshifter** [comment]
Does this need to be approved again now the conflict is fixed?

**Comment by bitshifter** [comment]
https://ci.appveyor.com/project/rust-lang/rust/build/1.0.6120/job/afn5undqdt9fsp0b/messages

> Build execution time has reached the maximum allowed time for your plan (180 minutes).

:'(

**Comment by xftroxgpx** [comment]
Can't compile rust locally anymore? (or I'm missing something?)
Any tips?

```bash
   Compiling alloc_system v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/liballoc_system)
   Compiling panic_unwind v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libpanic_unwind)
error[E0658]: the struct `#[repr(align(u16))]` attribute is experimental (see issue #33626)
  --> libstd/sync/mpsc/cache_aligned.rs:14:1
   |
14 | #[repr(align(64))]
   | ^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(repr_align)] to the crate attributes to enable

error: aborting due to previous error

error: Could not compile `std`.

Caused by:
  process didn't exit successfully: `/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/bootstrap/debug/rustc --crate-name std libstd/lib.rs --error-format json --crate-type dylib --crate-type rlib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 --cfg feature="alloc_jemalloc" --cfg feature="backtrace" --cfg feature="jemalloc" --cfg feature="panic-unwind" --cfg feature="panic_unwind" -C metadata=f7765955cd67cf2b -C extra-filename=-f7765955cd67cf2b --out-dir /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/release/deps --extern alloc_system=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/liballoc_system-c02aaefcdf3ed014.rlib --extern libc=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/liblibc-e8c93db2d9f7b1a1.rlib --extern panic_unwind=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libpanic_unwind-00b9a6cd0d4fc568.rlib --extern compiler_builtins=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libcompiler_builtins-f1b239799398f922.rlib --extern rustc_msan=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/librustc_msan-62258ea7aa2b71fe.rlib --extern alloc=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/liballoc-4d746674fccddbc7.rlib --extern panic_abort=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libpanic_abort-a0a3317b1f888bcd.rlib --extern rustc_tsan=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/librustc_tsan-dec54e9bf41c7f34.rlib --extern alloc_jemalloc=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/liballoc_jemalloc-2fbac0da8986f50b.rlib --extern std_unicode=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libstd_unicode-d98c7e52ed02392c.rlib --extern rustc_asan=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/librustc_asan-7dee8df74f43bae5.rlib --extern core=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libcore-1fa9134e95dd93e2.rlib --extern unwind=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libunwind-73b86850bc2ec62b.rlib --extern rustc_lsan=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/librustc_lsan-431a8de9030d5943.rlib -L native=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/native/libbacktrace/.libs -l static=backtrace -l dl -l rt -l pthread -L native=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/native/jemalloc/lib -L native=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/build/compiler_builtins-f322bdafb8005573/out` (exit code: 101)
thread 'main' panicked at 'command did not execute successfully: "/home/xftroxgpx/.cargo/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "--release" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libstd/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101', bootstrap/compile.rs:881:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failed to run: /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/bootstrap/debug/bootstrap build --incremental -j 5
Build completed unsuccessfully in 0:02:42

```


```
rustc 1.25.0-nightly (a0dcecff9 2018-01-24)
binary: rustc
commit-hash: a0dcecff90c45ad5d4eb60859e22bb3f1b03842a
commit-date: 2018-01-24
host: x86_64-unknown-linux-gnu
release: 1.25.0-nightly
LLVM version: 4.0
```

**Comment by kennytm** [comment]
@xftroxgpx What is your `config.toml`? If you have defined `[build] rustc`, comment it out and try again.

**Comment by xftroxgpx** [comment]
My [config.toml](https://github.com/xftroxgpx/a3/blob/7fb45a8db8b41a05566c52608e18ebf589b01c09/system/Z575/OSes/3archlinux/on_baremetal/filesystem_now/archlinux/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/config.toml#L1).

I had `rustc = "/home/xftroxgpx/.cargo/bin/rustc"`
<details><summary>Now, commented out, but still failed:</summary>

```bash
./x.py build --incremental -j 5
Updating submodules
downloading https://static.rust-lang.org/dist/2018-01-02/rust-std-beta-x86_64-unknown-linux-gnu.tar.gz
######################################################################## 100.0%
extracting /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/cache/2018-01-02/rust-std-beta-x86_64-unknown-linux-gnu.tar.gz
downloading https://static.rust-lang.org/dist/2018-01-02/rustc-beta-x86_64-unknown-linux-gnu.tar.gz
######################################################################## 100.0%
extracting /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/cache/2018-01-02/rustc-beta-x86_64-unknown-linux-gnu.tar.gz
   Compiling num-traits v0.1.41
   Compiling lazy_static v0.2.11
   Compiling getopts v0.2.15
   Compiling itoa v0.3.4
   Compiling unicode-xid v0.0.4
   Compiling dtoa v0.4.2
   Compiling quote v0.3.15
   Compiling cc v1.0.4
   Compiling serde v1.0.27
   Compiling libc v0.2.36
   Compiling cfg-if v0.1.2
   Compiling synom v0.11.3
   Compiling syn v0.11.11
   Compiling filetime v0.1.15
   Compiling num_cpus v1.8.0
   Compiling cmake v0.1.29
   Compiling build_helper v0.1.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/build_helper)
   Compiling serde_derive_internals v0.19.0
   Compiling serde_derive v1.0.27
   Compiling toml v0.4.5
   Compiling serde_json v1.0.9
   Compiling bootstrap v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/bootstrap)
    Finished dev [unoptimized] target(s) in 75.69 secs
Building stage0 std artifacts (x86_64-unknown-linux-gnu -> x86_64-unknown-linux-gnu)
   Compiling unwind v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libunwind)
   Compiling core v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libcore)
   Compiling cfg-if v0.1.2
   Compiling libc v0.2.36
   Compiling cc v1.0.4
   Compiling filetime v0.1.15
   Compiling build_helper v0.1.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/build_helper)
   Compiling std v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libstd)
   Compiling alloc_jemalloc v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/liballoc_jemalloc)
   Compiling compiler_builtins v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/rustc/compiler_builtins_shim)
   Compiling cmake v0.1.29
   Compiling rustc_tsan v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/librustc_tsan)
   Compiling rustc_lsan v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/librustc_lsan)
   Compiling rustc_asan v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/librustc_asan)
   Compiling rustc_msan v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/librustc_msan)
   Compiling libc v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/rustc/libc_shim)
   Compiling std_unicode v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libstd_unicode)
   Compiling panic_abort v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libpanic_abort)
   Compiling alloc v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/liballoc)
   Compiling panic_unwind v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libpanic_unwind)
   Compiling alloc_system v0.0.0 (file:///home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/liballoc_system)
error: the struct `#[repr(align(u16))]` attribute is experimental (see issue #33626)
  --> libstd/sync/mpsc/cache_aligned.rs:14:1
   |
14 | #[repr(align(64))]
   | ^^^^^^^^^^^^^^^^^^
   |
   = help: add #![feature(repr_align)] to the crate attributes to enable

error: aborting due to previous error

error: Could not compile `std`.

Caused by:
  process didn't exit successfully: `/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/bootstrap/debug/rustc --crate-name std libstd/lib.rs --error-format json --crate-type dylib --crate-type rlib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 --cfg feature="alloc_jemalloc" --cfg feature="backtrace" --cfg feature="jemalloc" --cfg feature="panic-unwind" --cfg feature="panic_unwind" -C metadata=2d4d16d24b0a1c7c -C extra-filename=-2d4d16d24b0a1c7c --out-dir /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps -L dependency=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/release/deps --extern std_unicode=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libstd_unicode-0d7cd53a0f8ca1cb.rlib --extern libc=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/liblibc-fb0532659c2770d1.rlib --extern rustc_lsan=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/librustc_lsan-49988d7fcfe221d9.rlib --extern panic_unwind=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libpanic_unwind-ceb6d2604463fa9a.rlib --extern alloc=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/liballoc-2fa1908321f62402.rlib --extern core=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libcore-81500df0e121824a.rlib --extern rustc_msan=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/librustc_msan-e792ac012c81d5de.rlib --extern compiler_builtins=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libcompiler_builtins-31f820ea1f446abe.rlib --extern alloc_system=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/liballoc_system-045f19d2833d64ad.rlib --extern rustc_asan=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/librustc_asan-ae57686ea748ec53.rlib --extern panic_abort=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libpanic_abort-df8ccc0bb4380645.rlib --extern unwind=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/libunwind-52a24300225aab4a.rlib --extern alloc_jemalloc=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/liballoc_jemalloc-c8af342068974a5a.rlib --extern rustc_tsan=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/deps/librustc_tsan-0794e094bc2cb159.rlib -L native=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/native/libbacktrace/.libs -l static=backtrace -l dl -l rt -l pthread -L native=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/native/jemalloc/lib -L native=/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/x86_64-unknown-linux-gnu/stage0-std/x86_64-unknown-linux-gnu/release/build/compiler_builtins-5d7fc88585475c99/out` (exit code: 101)
thread 'main' panicked at 'command did not execute successfully: "/home/xftroxgpx/.cargo/bin/cargo" "build" "--target" "x86_64-unknown-linux-gnu" "--release" "--features" "panic-unwind jemalloc backtrace" "--manifest-path" "/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/src/libstd/Cargo.toml" "--message-format" "json"
expected success, got: exit code: 101', bootstrap/compile.rs:881:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
failed to run: /home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build/bootstrap/debug/bootstrap build --incremental -j 5
Build completed unsuccessfully in 0:03:10

real	3m11.358s
user	5m35.554s
sys	0m18.022s

```

</details>
  

The one is using now (stage0) is:
```
!! Executing '/home/xftroxgpx/build/2nonpkgs/rust.stuff/rust/rust/build//x86_64-unknown-linux-gnu/stage0/bin//rustc' with args: '-vV'
rustc 1.24.0-beta.1 (5b496b726 2018-01-02)
binary: rustc
commit-hash: 5b496b726462e6d5fa92397e7bda991b903eb082
commit-date: 2018-01-02
host: x86_64-unknown-linux-gnu
release: 1.24.0-beta.1
LLVM version: 4.0
```

The one I was using was:
```
$ ~/.cargo/bin/rustc -vV
rustc 1.25.0-nightly (a0dcecff9 2018-01-24)
binary: rustc
commit-hash: a0dcecff90c45ad5d4eb60859e22bb3f1b03842a
commit-date: 2018-01-24
host: x86_64-unknown-linux-gnu
release: 1.25.0-nightly
LLVM version: 4.0
```

**Comment by kennytm** [comment]
@xftroxgpx Comment out `local-rebuild = true` as well?

**Comment by xftroxgpx** [comment]
> Comment out local-rebuild = true as well?

I think that was it, thank you @kennytm !
(still compiling)

**Pull request merged**